ls1intum/Artemis
# Automatically resume exams after logging in after session expiration

### Checklist

### Motivation and Context

The student should not continue to work when the session is no longer valid, but instead be required to log in again and work with a valid state. Locally saved changes should be loaded when they are more recent than what was saved on the server.

### Description

When the session is no longer valid and therefore saving the exam fails, the user is logged out and brought back to the login page. An error is displayed (see screenshots below) and the user has to log back in. Once logged back in the user is automatically redirected to the exam page.

### Steps for Testing

1. Create an exam and participate as student
2. Invalidate your session by e.g. altering your session token:
![grafik](https://user-images.githubusercontent.com/72132281/109295081-6dbf1480-782e-11eb-8942-2e522d6d1f2d.png)
(Do note that this didn't work for me on Firefox but I had to manually set the session timeouts in application-dev.yml.)
3. Attempt to save the exam or hand in
4. Make sure you are put back to the login page and an error is displayed (see screenshots below)
5. After the login you should be redirected to the exam not the courses page

Note that you won't necessarily be redirected to the exact exercise you had just edited when your session got invalid. Also make sure to test that test runs and normal exam runs work as usual. On a final note, the local state (things you entered while disconnected) should be preserved for all exercises except programming exercises.

### Screenshots

When the session expires and you get logged out
![grafik](https://user-images.githubusercontent.com/72132281/109294575-aca09a80-782d-11eb-9836-fe0eba9d3e73.png)
![grafik](https://user-images.githubusercontent.com/72132281/109294589-af02f480-782d-11eb-8eae-976f798067b5.png)
